File: subarray_equals_k.py
 # Time Complexity : O(n)
 # Space Complexity : O(n)
 # Did this code successfully run on Leetcode : Yes
 # Any problem you faced while coding this : None
class Solution:
    def subarraySum(self, nums: List[int], k: int) -> int:
        hashmap = {}
        hashmap[0] = 1
        a_sum = 0
        count = 0
        for i in range(len(nums)):
            a_sum += nums[i]
            if (a_sum-k) in hashmap:
                count += hashmap[a_sum-k]
            if a_sum not in hashmap:
                hashmap[a_sum] = 1
            else:
                hashmap[a_sum] = hashmap[a_sum] + 1
        return count

File: subarray_equals_k.cpp
//Time Complexity : O(n)  
// Space Complexity : O(n)  
// Did this code successfully run on Leetcode : Yes
// Any problem you faced while coding this : No

class Solution {
public:
    int subarraySum(vector<int>& nums, int k) {
        int n  = nums.size();
        
        int count=0;
        
        unordered_map<int, int> prevSum;
        int sum=0;
        
        for(int i=0;i<n;i++){
            sum+=nums[i];
            
            if(sum==k)
                count++;
            
            if(prevSum.find(sum-k) != prevSum.end())
                count+=prevSum[sum-k];
            
            prevSum[sum]++;
        }
        
        return count;
    }
};